### PR TITLE
ES6/7/8 Training

### PR DESCRIPTION
**Pull request by roybenenosh**

### CODE REVIEW

File: src/lab15/exercise1/index.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+function printTimes(options) {
+  const { text, times } = options;
+
+  for (let i = 0; i < times; i++) {
+    console.log(`${String(i + 1).padStart(2, '0')} ${text}`);
+  }
+}
+
+// print 'hello world' ten times:
+printTimes({ text: 'hello world', times: 10 });
```
Comment (ilyador) [inline src/lab15/exercise1/index.js:10]: Correct

File: src/lab15/exercise2/index.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+function printTimes(options) {
+  const { text, times } = options;
+
+  for (let i = 0; i < times; i++) {
+    console.log(`${String(i + 1).padStart(2, '0')} ${text}`);
+  }
+}
+
+// print 'hello world' five times:
+printTimes({ text: 'hello world', times: 5 });
```
Comment (ilyador) [inline src/lab15/exercise2/index.js:10]: How would you make it work if there was no `times` at all, like this
```js
printTimes({ text: 'hello world' });
```
but you would still want it to run 5 times by default?

File: src/lab15/exercise3/index.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  const { text, times } = options;
+
+  for (let i = 0; i < times; i++) {
+    console.log(`${String(i + 1).padStart(2, '0')} ${text}`);
+  }
+}
+
+function printTimesFromWeirdDevelopersDownstairs(weirdOptions) {
+  const { title, times } = weirdOptions;
+  const text = title;
```
Comment (ilyador) [inline src/lab15/exercise3/index.js:11]: There a special ES6 syntax that merges lines 10 and 11, try to find out:)

File: src/lab15/exercise4/index.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+function fib(n) {
+  if (n == 0 || n == 1) {
+    return 1;
+  }
+  return fib(n - 1) + fib(n - 2);
+}
+
+function timed(func) {
+  return function(args) {
```
Comment (ilyador) [inline src/lab15/exercise4/index.js]: Your solution is good, however, when you pass an argument here the way you do, you can pass only one argument to the timed function (unless you use an array or object)
There is an ES6 way that will allow you to pass any number of arguments.

File: src/lab15/exercise5/index.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+let arr1 = [10, 20, 30, 40];
+let arr2 = [10, 20, 30, 40];
+
+// first option  - extends the memory
+// second option - uses new memory
+arr1.push(50);
+arr2 = [...arr2, 50];
```
Comment (ilyador) [inline src/lab15/exercise5/index.js:7]: Correct!

File: src/lab4/exercise1/index.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+var container = document.querySelector('div');
+
+for (let i = 0; i < 4; i++) {
```
Comment (ilyador) [inline src/lab4/exercise1/index.js:3]: Correct!

File: src/lab4/exercise2/index.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+// the console will print an error, because we tried to access x inside a block before initializing x in this block
+let x = 10;
+console.log(x);
```
Comment (ilyador) [inline src/lab4/exercise2/index.js:3]: Correct!

File: src/lab4/exercise3/index.js [hunk 8]
```diff
@@ -0,0 +1,3 @@
+const a = [10, 20];
+Object.freeze(a);
```
Comment (ilyador) [inline src/lab4/exercise3/index.js:2]: Correct!

File: src/lab7/exercise3/index.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+class AutoBind {
+  constructor(btn) {}
+}
+
+class Counter extends AutoBind {
+  constructor(btn) {
+    super();
+    this.btn = btn;
+    this.btn.textContent = '0';
+    btn.addEventListener('click', this.inc.bind(this));
```
Comment (ilyador) [inline src/lab7/exercise3/index.js:10]: Your solution is working and correct.
Can you make it work without changing `Counter` and by making `AutoBind` generically do the bind operation for any method?

File: src/lab7/exercise2/index.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
+class Car {
+  constructor(color, score) {
+    this.color = color;
+    this.score = score;
+  }
+}
+
+class Race {
+  constructor(car1, car2) {
```
Comment (ilyador) [inline src/lab7/exercise2/index.js]: Can you think of a way to pass any number of cars to this function?

File: src/lab7/exercise2/index.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    this.car1 = car1;
+    this.car2 = car2;
+    this.winner = car1.score > car2.score ? car1 : car2;
```
Comment (ilyador) [inline src/lab7/exercise2/index.js]: What if I add a car later? How will I calculate a second race with a new car?

File: src/lab7/exercise1/lightBulbChild.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+let children = [];
+
+for (let index = 0; index < 4; index++) {
+  let parent = document.querySelector('.bulb');
+  let child = document.createElement('div');
+  child.id = index;
+  children.push(new LightBulbChild(parent, child));
+}
```
Comment (ilyador) [inline src/lab7/exercise1/lightBulbChild.js:36]: Good solution

**Review by ilyador — COMMENTED** [review]
Good job on fixing the PR.

File: src/lab15/exercise3/index.js (hunk 3 above)
Comment (roybenenosh) [reply src/lab15/exercise3/index.js:11]: I pushed most of the fixes except this. Still not sure what is the answer, after going through the videos.

Comment (roybenenosh) [reply src/lab15/exercise3/index.js:11]: I did not figure out the answer.

Comment (ilyador) [reply src/lab15/exercise3/index.js:11]: Here is a clue:)
https://codeburst.io/renaming-destructured-variables-in-es6-807549754972

Comment (roybenenosh) [reply src/lab15/exercise3/index.js:11]: Done ! thanks !

**Pull request merged**